orion/eval/evaluation.py

```diff
@@ -0,0 +1,241 @@
+import numpy as np
+from sklearn import metrics
+
+
+class Evaluator:
+    """Evaluator Class.
+
+    The Evaluator Class provides a mechanic to assess the perfomance of
+    anomaly detection methods. This class addresses the problem as
+    contextual anomalies.
+
+    anomaly = [(start, end)] inclusive interval.
+
+    """
+
+    def _overlap(self, expected, observed):
+        first = expected[0] - observed[1]
+        second = expected[1] - observed[0]
+        return first * second < 0
+
+    def _any_overlap(self, part, intervals):
+        for interval in intervals:
+            if self._overlap(part, interval):
+                return 1
+
+        return 0
+
+    def _pad(self, lst):
+        return [(part[0], part[1] + 1) for part in lst]
+
+    def _partition(self, expected, observed, start=None, end=None):
+        edges = set()
+
+        if start is not None:
+            edges.add(start)
+
+        if end is not None:
+            edges.add(end)
+
+        for edge in expected + observed:
+
+            edges.update(edge)
+
+        partitions = list()
+        edges = sorted(edges)
+        last = edges[0]
+        for edge in edges[1:]:
+            partitions.append((last, edge))
+            last = edge
+
+        expected_parts = list()
+        observed_parts = list()
+        weights = list()
+        for part in partitions:
+            weights.append(part[1] - part[0])
+            expected_parts.append(self._any_overlap(part, expected))
+            observed_parts.append(self._any_overlap(part, observed))
+
+        return expected_parts, observed_parts, weights
+
+    def _weighted_segment(self, expected, observed, start=None, end=None):
+        expected, observed, weights = self._partition(expected, observed, start, end)
+
+        return metrics.confusion_matrix(
+            expected, observed, sample_weight=weights, labels=[0, 1]).ravel()
+
+    def _overlap_segment(self, expected, observed, start=None, end=None):
+        tp, fp, fn = 0, 0, 0
+
+        observed_copy = observed.copy()
+
+        for expected_seq in expected:
+            found = False
+            for observed_seq in observed:
+                if self._overlap(expected_seq, observed_seq):
+                    if not found:
+                        tp += 1
+                        found = True
+                    if observed_seq in observed_copy:
+                        observed_copy.remove(observed_seq)
+            if not found:
+                fn += 1
+        fp += len(observed_copy)
+
+        return None, fp, fn, tp
+
+    def get_confusion_matrix(self, expected, observed, method=None,
+                             data=None, start=None, end=None):
+        """Compute the confusion matrix between the ground truth and the detected anomalies.
+
+        Args:
+            * expected (pd.DataFrame): Ground truth.
```

Comment:
A couple of things should be considered regarding the docstrings:
- The `*` can be removed when listing the `Args`, so line starts directly with the argument name.
- A `Returns` block should be added listing the return type and a description in the next line.
- A blank line should always exist between the first line and the rest, and also before Args and Returns
- (Optional): I particularly like to put the argument description in a line below the argument name instead of putting all of it in the same line. This ensures that format is properly parsed when documentation is rendered on Sphinx if the description has more than one line.

```
"""Short description in a single line ending with a dot.

Longer description that can span across multiple lines. Longer
description that can span across multiple lines. Longer description
that can span across multiple lines.

Args:
    arg_name (arg_type):
        argument description.
    arg_name (arg_type):
        Argument description that spans across multiple lines. Argument
        description that spans across multiple lines. Argument description
        that spans across multiple lines.

Returns:
    return_type:
        description of the returned objects
```